### ISSUE TITLE
Jetson camera IMX219-160 → MNIST inference demo

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#396]
## Goal

Drive the IMX219-160 CSI camera on the Jetson Orin Nano dev kit from SLM-OS, capture a frame, preprocess it to MNIST input format, and run the captured digit through the existing MNIST pipeline — end-to-end from Lua.

Target demo:
```lua
local cam = slm.camera.open("imx219-0")
local frame = cam:capture()
local mnist = slm.camera.preprocess_mnist(frame)
local logits, pred = slm.model_infer_bytes(mnist_idx, mnist)
slm.print(string.format("predicted digit: %d", pred))
cam:close()
```

Detailed plan: [`docs/jetson-camera-imx219-plan.md`](https://github.com/SLM-OS/SLM-Operating-System/blob/main/docs/jetson-camera-imx219-plan.md).

## Why this is feasible

- **Inference half is already done.** `slm.model_infer_bytes(idx, fp32_bytes)` (commit `886dd0d`, PR #386) executes the MNIST CNN on the GA10B GPU when a v6 handoff is loaded, with a CPU fallback otherwise. The camera path just needs to produce 28×28×fp32 = 3 136 bytes.
- **BPMP MRQ stack works post-kexec** (commits `1f261e2`, `0f24e48`, `ea28cdc`). `bpmp_clk_enable` / `bpmp_reset_deassert` / `bpmp_pg_set_state` are functional and have been used to bring up PCIe-C8 (#25 Step 3). Clock, reset, and power-domain control for I2C, NVCSI, and VI are reachable.

## What's new (greenfield)

- Tegra HSI2C driver (no I²C bus master in the repo today).
- IMX219 sensor driver (I²C-side state machine + mode tables; full Linux reference exists).
- NVCSI driver (Tegra234 MIPI CSI-2 receiver — undocumented; reverse-engineer from L4T sources).
- VI driver (Video Input + DMA — even less documented).
- Image preprocessing pipeline (Bayer crop → grayscale → 28×28 → fp32 normalize).
- Lua bindings + demo script.

## Single biggest unknown

**CBB firewall posture for NVCSI / VI / camera I²C at NS EL2.** Pattern from `docs/jetson-cbb-report.md`: peripherals Linux uses at NS EL1 are *usually* inherited at NS EL2-VHE, but there are exceptions (UARTA is blocked despite Linux using it). NVCSI/VI/I²C are "regular" peripherals (no secure-only role), so this is a coin flip. Must be probed on hardware before committing.

The plan opens with a **Phase 0 hardware recon** (1–2 days) — single-shot `mem peek` of NVCSI and VI MMIO bases. If CBB blocks either, the bare-metal path is not viable from NS EL2 and one of three fallbacks applies (Linux pre-kexec snapshot, USB UVC camera, or skipping live capture and demo'ing the embedded test digits already present at `/mnt/files/digits/digit_N.bin`).

## Effort

| Phase | Range |
|-------|-------|
| Phase 0 (CBB recon) | 1–2 d |
| Tegra HSI2C driver | 4–6 d |
| IMX219 sensor driver | 3–5 d |
| NVCSI driver | 7–12 d |
| VI driver (single-shot) | 8–12 d |
| Image preprocessing | 2–3 d |
| Lua bindings + demo | 1.5–3 d |
| Hardware iteration buffer | 4–8 d |
| QEMU-side regression tests | 2–3 d |
| **Total (happy path)** | **32–54 d (~7–11 weeks)** |

This is a **multi-month commitment** comparable in scope to the GA10B GPU compute bring-up (#356) or the dynamic-kernel-replace plan (#369-family). For the capstone milestone it is plausibly **post-capstone** unless triaged otherwise.

## Recommended sequencing

1. **Land Phase 0 first** as a 1–2 day investigation. The answer to "can SLM-OS at EL2 even read NVCSI and VI?" decides whether to proceed.
2. If green: prioritize **QEMU-side mock-camera + preprocessing** first. That delivers a runnable demo on QEMU within a week and de-risks the Lua / inference plumbing before any Tegra reverse-engineering work.
3. **Tegra HSI2C + IMX219** next — verifiable in isolation (CHIP_ID readback at register `0x0000` returns `0x0219`).
4. **NVCSI + VI** last, behind iteration buffer.

## Risks (from the plan)

| # | Risk | Mitigation |
|---|------|-----------|
| 1 | CBB firewall blocks NVCSI/VI/I²C | Phase 0 recon; fall back to Linux-snapshot or USB-UVC if blocked |
| 2 | NVCSI/VI undocumented + RTCPU dependency | Code-read L4T sources before committing; RTCPU IVC adds 1–2 weeks if required |
| 3 | SMMU drops VI stream-id translations | Code-read DT for VI's `iommus` property + SMMU bypass test; full SMMU programming is a separate large work item if needed |
| 4 | No IMX219 module in lab | Confirm with maintainer; ~$25 + a few days lead time if absent |
| 5 | Linux idle gates camera clocks at kexec | Use BPMP to enable clocks ourselves, or pre-kexec hold via a script (same pattern as `slmos-kexec`) |

## Out of scope

- **Continuous capture / streaming.** Single-shot capture only; ring-buffer adds ~1 week.
- **ISP pipeline / 3A (auto-exposure, auto-white-balance, auto-focus).** Raw Bayer → preprocess → MNIST is enough for digit recognition.
- **Multi-camera.** Single sensor on one CSI port.
- **HDR or higher-bit-depth modes.** RAW10 only.

## Related

- #25 — Jetson NIC (BPMP IPC) — same BPMP infrastructure being reused.
- #266 — Jetson USB networking — provides the SMMU-at-kexec lessons that VI DMA will probably hit.
- #356 — Jetson GA10B compute kernel launch — the GPU side of the inference path the camera feeds into.
- `docs/jetson-cbb-report.md` — CBB firewall envelope.
- `docs/jetson-bpmp-ipc-plan.md` — BPMP MRQ infrastructure.

**Comment by johnjezl** [comment]
## Updated agent prompt — `make test` baseline as of 2026-04-26

Yesterday's prompt (and the earlier "45-failure trap" wording from PR #365 / #398) referenced pre-existing `make test` failures on `origin/main`. As of `01921c8` (PR #402, `hotfix/main-test-repair-20260425`), **`origin/main` is fully green again** — `make test` exits 0, ~1 849 PASS lines, zero failures.

The "ignore the baseline failures" caveat is therefore obsolete. Replace the relevant `VERIFY YOUR WORK:` block in the agent prompt with:

```
VERIFY YOUR WORK:

- Use `/ci` for the local CI sweep.
- Use `/slmos-review` for the SLM-OS-tailored review pass before opening a PR.
- Post-change checklist (per kernel/CLAUDE.md): tests + docs + make test pass
  before reporting completion.
- `make test` on origin/main is fully green as of 2026-04-26 (exit 0,
  ~1849 [PASS], 0 [FAIL]). Any new failure your branch introduces is a
  real regression — not pre-existing baseline noise. Always re-run on a
  fresh `origin/main` rebase before assuming a failure is yours.
```

For reference, the failures that existed yesterday and were swept up by `#402`:

| Yesterday's failure | Today |
|---|---|
| `trajectory_records_feedback_events` | ✅ PASS |
| `test_eviction_run_tests_passes` | ✅ PASS |
| `test_migrate_ready_task_updates_assigned_cpu` | ✅ PASS |
| `test_parse_http_url_accepts_long_signed_uri` | ✅ PASS |
| `test_start_and_enumerate` | ✅ PASS |
| `test_descriptor_parse` | ✅ PASS |
| `test_find_endpoint_mismatch` | ✅ PASS |

If a future `main` regresses again, drop a fresh comment here with the new baseline so the agent prompt stays accurate.

**Comment by johnjezl** [comment]
Closed by PR #818 — end-to-end IMX219 → MNIST inference demo working from a single Lua entry point.

## Hardware verification (jetson-nano-1, 2026-05-14)

`lua-admin /mnt/files/demo_imx219_mnist.lua` after `slmos-kexec`:

```
1. Capture — slm.camera.open() + cam:capture()
   camera.open: 390 ms  (first call — runs full RCE/IVC bring-up + IMX219 power-on)
   capture:     188 ms  (frame_id=0, 1640x1232, bayer=0)

2. Preprocess — slm.camera.preprocess_mnist()
   preprocess:  191 ms  (3136 bytes = 784 fp32 samples)

3. Model — slm.model_load_mnist()
   model_load:    0 ms  (model_id=1, mnist preloaded at boot)

4. Inference — slm.model_infer_bytes()
   inference:     4 ms

   Logits (class -> score):
     0:  -0.0506
     1:   0.0141
     2:   0.0519
     3:   0.0169
     4:  -0.1127
     5:   0.1495  <-- argmax
     6:  -0.0681
     7:  -0.0597
     8:   0.0955
     9:  -0.0378

   Predicted digit: 5
```

The argmax has no semantic meaning until the camera is pointed at a hand-drawn digit on a uniform background. The verification confirms the full pipeline produces a valid logits vector end-to-end.

## What landed (chronological)

- **#812** — IMX219 LP-11 force at power-on (step 1 of the issue)
- **#813** — IMX219 first light (closed the three root causes blocking #518: NVCSI port B not A, D-PHY CIL `lp_bypass_mode=1` + `lane_polarity`, EMBDATA channel-flag + per-frame fields + surface routing)
- **#814** — `camera_open("imx219-0")` backend in `camera.c`; refactored `csidiag` to a thin wrapper over the same path
- **#816** — Optional `camera-name` 5th arg on `slm.camera.preprocess_mnist` so Lua callers can target the real sensor
- **#818** — `scripts/demo_imx219_mnist.lua` + embedded-scripts wiring; hardware-verified end-to-end on jetson-nano-1

## Phase 0 recon outcome (recap)

The "single biggest unknown" from the original plan — CBB firewall posture at NS EL2 for NVCSI / VI / camera I²C — resolved as: **NVCSI direct MMIO is blocked from any AP context (the Camera RTCPU IVC path is the only route), but HSI2C and VI work at EL2 via RTCPU's capture-control + capture channels.** The implementation routes everything through `kernel/drivers/camrtc/` (RCE HSP-VM + IVC), not direct NVCSI MMIO.

## Follow-ups

- **#815** — Refactor `camera_preprocess_mnist` into composable Rust-side preprocessing before the next model lands (the current shape is MNIST-specific: 1640×1232 RGGB → green-only → 28×28 fp32).
- **#466** — `camrtc` commit/poll lock split (pre-existing P3).

## Out of scope (kept as in the original issue)

Continuous capture / streaming, ISP / 3A, multi-camera, and HDR all remain unimplemented and untracked.
